### ISSUE TITLE
Spatial softmax for segmentation

### ISSUE THREAD

**Issue by bluechill52** [vlfeat/matconvnet#269]
@vedaldi Hi, I've just started using matconvnet and have got my hands dirty with cifar and mnist. I am trying to do semantic segmentation of an image. From the vl_nnsoftmaxloss equation

Y = VL_NNSOFTMAX(X, C)

I get that the labels can be of size H x W x 1 x N. I am a little confused as to what is N here. For example, if my batch size is 1, then should the size of X be H x W x D x 1 and labels be H x W respectively ? And in that case, what is the size of the output Y ? Thanks in advance.

**Comment by dasguptar** [comment]
`vl_nnsoftmaxloss` is no longer supported. The convention now is to use `vl_softmax` followed by `vl_nnloss`, where the latter now has multiple types of loss functions supported.
`vl_nnloss`  takes `X` and `c` as inputs, where `X` needs to be of size `H x W x D x N`.
For segmentation, each spatial location would require a corresponding ground truth, so the size of `c` needs to be `H x W x 1 x N`, where `N` is the batch size.
So you are correct in assuming that for a batch size of  1, `X` will be `H x W x D x 1`, and `c` will be `H x W x 1 x 1`. Each location in `c` is to contain the class label for semantic segmentation. 
To get the dimensionality of `Y`, go through the documentation at https://github.com/vlfeat/matconvnet/blob/master/matlab/vl_nnloss.m#L13
